Readme.md
==========
b8: readme
==========

:Author: Tobias Leupold
:Homepage: http://nasauber.de/
:Contact: tobias.leupold@web.de
:Date: @LASTCHANGE@

.. contents:: Table of Contents

Description of b8
=================

What is b8?
-----------

b8 is a spam filter implemented in `PHP <http://www.php.net/>`__. It is intended to keep your weblog or guestbook spam-free. The filter can be used anywhere in your PHP code and tells you whether a text is spam or not, using statistical text analysis. What it does is: you give b8 a text and it returns a value between 0 and 1, saying it's ham when it's near 0 and saying it's spam when it's near 1. See `How does it work?`_ for details about this. |br|
To be able to do this, b8 first has to learn some spam and some ham (non-spam) texts. If it makes mistakes when classifying unknown texts or the result is not distinct enough, b8 can be told what the text actually is, getting better with each learned text.

b8 is a statistical spam filter. I'm not a mathematician, but as far as I can grasp it, the math used in b8 has not much to do with Bayes' theorem itself. So I call it a *statistical* spam filter, not a *Bayesian* one. Principally, It's a program like `Bogofilter <http://bogofilter.sourceforge.net/>`__ or `SpamBayes <http://spambayes.sourceforge.net/>`__, but it is not intended to classify emails. Therefore, the way b8 works is slightly different from email spam filters. See `What's different?`_ if you're interested in the details.

An example of what we're talking about here:

At the moment of this writing (november 2012), b8 has, since december 2006, classified 26869 guestbook entries and weblog comments on my homepage. 145 were ham. 76 spam texts (0.28 %) have been falsely rated as ham (false negatives) and I had to remove them manually. Only one single ham message has been falsely classified as spam (false positive) back in june 2010, but – in defense of b8 – this was the very first English ham text I got. Previously, each and every of the 15024 English texts posted has been spam. Texts with Chinese, Japanese or Cyrillic content (all spam either) did not appear until 2011. |br|
This results in a sensitivity of 99.72 % (the probability that a spam text will actually be rated as spam) and a specifity of 99.31 % (the probability that a ham text will actually be rated as ham) for my homepage. Before the one false positive, of course, the specifity has been 100 % ;-)

How does it work?
-----------------

In principle, b8 uses the math and technique described in Gary Robinson's articles "A Statistical Approach to the Spam Problem" [#statisticalapproach]_ and "Spam Detection" [#spamdetection]_. The "degeneration" method Paul Graham proposed in "Better Bayesian Filtering" [#betterbayesian]_ has also been implemented.

b8 cuts the text to classify to pieces, extracting stuff like email addresses, links and HTML tags and of course normal words. For each such token, it calculates a single probability for a text containing it being spam, based on what the filter has learned so far. When the token has not been seen before, b8 tries to find similar ones using "degeneration" and uses the most relevant value found. If really nothing is found, b8 assumes a default rating for this token for the further calculations. |br|
Then, b8 takes the most relevant values (which have a rating far from 0.5, which would mean we don't know what it is) and calculates the combined probability that the whole text is spam.

What do I need for it?
----------------------

Not much! You just need PHP 5 and a database to store the wordlist.

Currently, there are three backends to choose from for use with one of the following databases:

* `Berkeley DB <http://oracle.com/technetwork/products/berkeleydb/downloads/index.html>`_
* `MySQL <http://mysql.com/>`_
* `PostgreSQL <http://postgresql.org/>`_

What's different?
-----------------

b8 has been designed to classify forum posts, weblog comments or guestbook entries, not emails. For this reason, it uses a slightly different technique than most of the other statistical spam filters out there use.

My experience was that spam entries on my weblog or guestbook were often quite short, sometimes just something like "123abc" as text and a link to a suspect homepage. Some spam bots don't even made a difference between e. g. the "name" and "text" fields and posted their text as email address, for example. Considering this, b8 just takes one string to classify, making no difference between "headers" and "text". |br|
The other thing is that most statistical spam filters count one token one time, no matter how often it appears in the text (as Paul Graham describes it in [#planforspam]_). b8 does count how often a token has been seen and learns resp. considers this. Why this? Because a text containing one link (no matter where it points to, just indicated by a "\h\t\t\p\:\/\/" or a "www.") might not be spam, but a text containing 20 links might be.

This means that b8 might be good for classifying weblog comments, guestbook entries or forum posts (I really think it is ;-) – but very likely, it will work quite poor when being used for something else like classifying emails. At least with the default lexer. But as said above, for this task, there are lots of very good filters out there to choose from.


Installation
============

Installing b8 on your server is quite easy. You just have to provide the needed files. To do this, you could just upload the whole ``b8`` subdirectory to the base directory of your homepage. It contains the filter itself and all needed backend classes. The other directories (``doc``, ``example``, ``install`` and ``update``) are not used by b8.

That's it ;-)

Configuration
=============

The configuration is passed as arrays when instantiating a new b8 object. Four arrays can be passed to b8. One containing b8's base configuration, one for the storage backend, one for the lexer and one for the degenerator. |br|
You can have a look at ``example/index.php`` to see how this can be done. `Using b8 in your scripts`_ also shows example code showing how b8 can be included in a PHP script.

Not all values have to be set. When some values are missing, the default ones will be used. If you do use the default settings, you don't have to pass them to b8. But of course, if you want to set something in e. g. the fourth config array, but not in the third, you will have to pass an empty ``array()`` as third parameter anyway.

b8's base configuration
-----------------------

All these values can be set in the "config_b8" array (the first parameter) passed to b8. The name of the array doesn't matter (of course), it just has to be the first argument.

These are some basic settings telling b8 which backend classes to use:

    **storage**
        This defines which storage backend will be used to save b8's wordlist. Currently, three databases are supported: `Berkeley DB <http://oracle.com/technetwork/products/berkeleydb/downloads/index.html>`_ (``dba``), `MySQL <http://mysql.com/>`_ (``mysql`` and ``mysqli``) and `PostgreSQL <http://postgresql.org/>`_ (``postgresql``). An experimental backend for `SQLite <http://sqlite.org/>`_ resides in SVN trunk but has not reached release quality yet. The default is ``dba`` (string).

        *dba (Berkeley DB)*
            This has been the original backend for the filter. All content is saved in a single file, you don't need special user rights or a database server. Probably a good choice, as this is very performant and fits exactly to b8's needs. |br|
            If you don't know whether your server's PHP installation supports Berkeley DB, simply run the script ``install/setup_berkeleydb.php``. If it shows a Berkeley DB handler, you can use this backend.

        *mysqli (MySQL)*
            The MySQL relational database system is used very widely on the web and can also be used for storing b8's wordlist. This backend needs of course a running and accessable MySQL server and database. The backend uses the mysqli\_* PHP functions to interact with the database.

        *mysql (MySQL)*
            This is the original MySQL backend using the legacy mysql\_* PHP functions. PHP encourages users to use the newer mysqli\_* functions. The functions used by this backend have been removed in PHP 7.0.0, so eventually, this backend will also be removed from b8.

        *postgresql (PostgreSQL)*
            A PostgreSQL schema with one table can also be used for storing b8's wordlist. This backend needs of course a running and accessable PostgreSQL server and database. |br|
            Communication with the DB server is done via PDO, so you need PHP >= 5.1 compiled with ``--with-pdo-pgsql`` to use this backend.

        See `Configuration of the storage backend`_ for the settings of the chosen backend.

    **lexer**
        The lexer class to be used. Defaults to ``default`` (string). |br|
        At the moment, only one lexer exists, so you probably don't want to change this unless you have written your own lexer.

    **degenerator**
        The degenerator class to be used. See `How does it work?`_ and [#betterbayesian]_ if you're interested in what "degeneration" is. Defaults to ``default`` (string). |br|
        At the moment, only one degenerator exists, so you probably don't want to change this unless you have written your own degenerator.

The following settings influence the mathematical internals of b8. If you want to experiment, feel free to play around with them; but be warned: wrong settings of these values will result in poor performance or could even "short-circuit" the filter. Leave these values as they are unless you know what you are doing.

The "Statistical discussion about b8" [#b8statistic]_ shows why the default values are the default ones.

    **use_relevant**
        This tells b8 how many tokens should be used to calculate the spamminess of a text. The default setting is ``15`` (integer). This seems to be a quite reasonable value. When using too many tokens, the filter will fail on texts filled with useless stuff or with passages from a newspaper, etc. not being very spammish. |br|
        The tokens counted multiple times (see above) are added in addition to this value. They don't replace other interesting tokens.

    **min_dev**
        This defines a minimum deviation from 0.5 that a token's rating must have to be considered when calculating the spamminess. Tokens with a rating closer to 0.5 than this value will simply be skipped. |br|
        If you don't want to use this feature, set this to ``0``. Defaults to ``0.2`` (float). Read [#b8statistic]_ before increasing this.

    **rob_x**
        This is Gary Robinson's *x* constant (cf. [#spamdetection]_). A completely unknown token will be rated with the value of ``rob_x``. The default ``0.5`` (float) seems to be quite reasonable, as we can't say if a token that also can't be rated by degeneration is good or bad. |br|
        If you receive much more spam than ham or vice versa, you could change this setting accordingly.

    **rob_s**
        This is Gary Robinson's *s* constant. This is essentially the probability that the *rob_x* value is correct for a completely unknown token. It will also shift the probability of rarely seen tokens towards this value. The default is ``0.3`` (float) |br|
        See [#spamdetection]_ for a closer description of the *s* constant and read [#b8statistic]_ for specific information about this constant in b8's algorithms.


Configuration of the lexer
--------------------------

The lexer disassembles the text we want to analyze to single words ("tokens"). The way it does this can be customized.

All the following values can be set in the "config_lexer" array (the third parameter) passed to b8. The name of the array doesn't matter (of course), it just has to be the third argument.

**min_size**
    The minimal length for a token to be considered when calculating the rating of a text. Defaults to ``3`` (integer).

**max_size**
    The maximal length for a token to be considered when calculating the rating of a text. Defaults to ``30`` (integer).

**allow_numbers**
    Should pure numbers also be considered? Defaults to ``false`` (boolean).

**get_uris**
    Look for URIs. Defaults to ``true`` (boolean).

**old_get_html**
    Extracts HTML. This is the old search function used up to b8 0.5.2. If you have an existing b8 installation and want the exactly same behaviour as before, use this. This function will probably removed in a future release. Defaults to ``true`` (boolean).

**get_html**
    Extracts HTML. This has been added in b8 0.6 and should work better then the "old_get_html" procedure. Defaults to ``false`` (boolean).

**get_bbcode**
    Extracts BBCode, which is often used in forums. Defaults to ``false`` (boolean).

Configuration of the degenerator
--------------------------------

When a token is not found in the database, b8 tries to find similar versions of that token. The degenerator provides these similar versions (cf. [#betterbayesian]_). The way it generates these "degenerates" can be customized.

All the following values can be set in the "config_degenerator" array (the fourth parameter) passed to b8. The name of the array doesn't matter (of course), it just has to be the fourth argument.

**multibyte**
    Use multibyte operations when searching for degenerated versions of an unknown token. When activating this, b8 needs PHP's ``mbstring`` module to work. Defaults to ``false`` (boolean).

**encoding**
    The internal encoding to use when doing multibyte operations. This will only be used when ``multibyte`` is set to ``true``. Defaults to ``UTF-8`` (string).

The difference of using or not using multibyte operations will only show up when non-latin-1 text is processed by b8. For example, if we have an unknown token ``HeLlO!``, the degenerator will provide the degenerated versions ``hello!``, ``HELLO!``, ``Hello!``, ``hello``, ``HELLO``, ``Hello`` and ``HeLlO``, no matter if multibyte operations are used or not.

When we have a non-latin-1 word, we may get a different result. For example, if we have the unknow token ``ПрИвЕт!``, the degenerator will only provide one degenerated version of it: ``ПрИвЕт``. Using multibyte operations, we get the same variants as with the latin-1 word: ``привет!``, ``ПРИВЕТ!``, ``Привет!``, ``привет``, ``ПРИВЕТ``, ``Привет`` and ``ПрИвЕт``.

Using multibyte operations will simply make the degenerator more effective.

Using b8
========

Now, that everything is configured, you can start to use b8. A sample script that shows what can be done with the filter can be found in ``example/``. Using this script, you can test how all this works before integrating b8 in your own scripts.

Before you can start, you have to setup a database so that b8 can store a wordlist.

Setting up a new database
-------------------------

Setting up a new Berkeley DB
````````````````````````````

There's a script that automates setting up a new Berkeley DB for b8. It is located at ``install/setup_berkeleydb.php``. Just run this script on your server and be sure that the directory containing it has the proper access rights set so that the server's HTTP server user or PHP user can create a new file in it (probably ``0777``). The script is quite self-explaining, just run it.


CHANGELOG
2019-10-15 Joao Gilberto Magalhaes <github.com/byjg>

    * Release: Version 0.7.0

    * Full refactoring of to code to use a ORM as Storage and creating interfaces for Lexer and Degenerator.

    * Library can now be imported using composer.

    * Dropped support from PHP 5.5 or less.

    * Added Namespace to the classes

2019-02-08 Tobias Leupold <tobias.leupold@web.de>

    * Release: Version 0.6.2

    * b8.php: Removed the use of the (as of PHP 7.2) deprecated each() function. Thanks to Markus Hauser <mh@diebeiden.at> for the bug report along with a proposed patch!

    * Overall code cleanup: Fixed coding style, indentation, function names, variable names ...

2014-03-12 Tobias Leupold <tobias.leupold@web.de>

    * Release: Version 0.6.1

    * storage/storage_postgresql.php: Added a backend for PostgreSQL, written by Tom Regner <tom@goochesa.de>

    * storage/storage_mysql.php: Fixed the theoretical possibility for an SQL injection. Thanks to Dirk Stolle for the bug report!

    * storage/storage_mysqli.php: Added a MySQL backend using the new mysqli_* functions instead of the legacy mysql_* ones. Thanks to Lorenzo Masetti <lorenzo.masetti@gmail.com> for the initial port!

2012-11-30 Tobias Leupold <tobias.leupold@web.de>

    * Release: Version 0.6

    * Lots of changes (new major release!):
        - Finally did an actually really complete abstraction of the storage backends, so that it was possible to change MySQL's database layout to multiple columns.
        - Kicked out the never-used lastseen parameter.
        - Renamed the internal variables to "b8*...", combined "bayes*text.ham" and "bayes*texts.spam" to "b8*texts".
        - Thus, the database layout has been changed. Included update scripts for both MySQL and DBA.
        - Removed all validate() functions in favor of throwing exceptions when something's wrong.
        - Made the lexer more flexible. Added functions for all split work, that, except for the raw split, can be turned on and off via a new config array.
        - The lexer now supports getting BBCode.
        - Added an additional check to the lexer to be sure no token will collide with an internal variable
        - Added multibyte support to the degenerator so it is now able to handle non-latin-1 texts in the same way as it handles latin-1-texts.
        - b8's constructor now takes four config arrays, the third is the lexer config, the fourth is the degenerator config.

    * Bugfixes:
        - Removed the ucfirst function from the degenerator and replaced it with a custom one. It did not what I always thought it would do (first letter upper case, rest lower case), but does only converted the first letter to upper case.
        - Fixed the MySQL backend so it's now able to handle a get() request for an empty array or an array containing just one token.
        - Fixed the MySQL backend when doing a query with no returned result.
        - Fixed the lexer to never output an empty array of tokens, but a placeholder token if no token has been found.

2011-04-19 Tobias Leupold <tobias.leupold@web.de>

    * Release: Version 0.5.2

    * storage/storage_mysql.php: The destructor of the MySQL storage backend always disconnected the MySQL connection, even if an existing link-resource was passed to b8. Now, the MySQL connection is only disconnected if b8 created the resource itself. Thanks to Gerhard Waldemair for the bug report!

2010-12-30 Tobias Leupold <tobias.leupold@web.de>

    * Release: Version 0.5.1

    * Bigger changes:
        - Fixed some issues with the scope of variables leading to problems when multiple instances of b8 are created. Thanks to Mike Creuzer for the bug report :-)
        - Centralized the loading of class definition files in the b8 constructor and created a function to handle the inclusion.

    * b8.php: Return a lexer error code instead of a rating if the lexer failed. The lexer never returned FALSE but b8 checked only for this value to validate the lexer didn't fail. Thanks to Matt Friedman for the bug report :-)

    * lexer/lexer_default.php: A bit of code cleanup: less useless nesting.

    * doc/readme.*: Updated the documentation, added a FAQ.

2010-06-27 Tobias Leupold <tobias.leupold@web.de>

    * Release: Version 0.5-r1

    * doc/readme.*: Updated the documentation; forgot the newly introduced b8::HAM and b8::SPAM variables. Added some additional information about the storage model.

2010-06-02 Tobias Leupold <tobias.leupold@web.de>

    * Release: Version 0.5

    * 100.000 Changes (new major release!), at a glance:
        - No PHP 4 compatibility anymore. Much cleaner code base with less hacks.
        - Completely reworked storage model. The SQL performance increased dramatically, the Berkeley DB performance remains as fast as it always has been.
        - Better lexer which can also handle non-latin1 texts in a nice way, so that e. g. Cyrillic or Chinese texts can be classified more performant.
        - No config files anymore, multiple instances of b8 can be now created in the same script with different configuration, databases and no problems.
        - No spooky administration interface anymore that needs an SQL database, even if Berkeley DB is used (anybody who actually used this?! I never did ;-).
        - No "install" scripts and routines and a less end-user compatible documentation. Anybody integrating b8 in his homepage won't be an end-user, will he?

2009-02-03 Oliver Lillie (aka buggedcom)

    * Revision: 221 (the original PHP 5 port)

    * Rewrote Tobias' original class for optimisation and PHP 5 functionality.

    * Improved database mysql query useage by over ~820%

    * Class is faster, ~20%.

    * Slight increase in memory usage, but it's small and given the advantages of the speed increase and query reduction it's worth it.


b8/b8.php
<?php

#   Copyright (C) 2006-2018 Tobias Leupold <tobias.leupold@web.de>
#
#   b8 - A statistical ("Bayesian") spam filter written in PHP 5
#
#   This program is free software; you can redistribute it and/or modify it
#   under the terms of the GNU Lesser General Public License as published by
#   the Free Software Foundation in version 2.1 of the License.
#
#   This program is distributed in the hope that it will be useful, but
#   WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
#   or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
#   License for more details.
#
#   You should have received a copy of the GNU Lesser General Public License
#   along with this program; if not, write to the Free Software Foundation,
#   Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

/**
 * Copyright (C) 2006-2018 Tobias Leupold <tobias.leupold@web.de>
 *
 * @license LGPL 2.1
 * @access public
 * @package b8
 * @author Tobias Leupold
 * @author Oliver Lillie (original PHP 5 port)
 */

class b8
{

    const DBVERSION = 3;

    public $config = array(
        'lexer'        => 'default',
        'degenerator'  => 'default',
        'storage'      => 'dba',
        'use_relevant' => 15,
        'min_dev'      => 0.2,
        'rob_s'        => 0.3,
        'rob_x'        => 0.5
    );

    public $storage     = null;
    public $lexer       = null;
    public $degenerator = null;

    private $_token_data = null;

    const SPAM    = 'spam';
    const HAM     = 'ham';
    const LEARN   = 'learn';
    const UNLEARN = 'unlearn';

    const CLASSIFYER_TEXT_MISSING = 'CLASSIFYER_TEXT_MISSING';

    const TRAINER_TEXT_MISSING     = 'TRAINER_TEXT_MISSING';
    const TRAINER_CATEGORY_MISSING = 'TRAINER_CATEGORY_MISSING';
    const TRAINER_CATEGORY_FAIL    = 'TRAINER_CATEGORY_FAIL';

    /**
     * Constructs b8
     *
     * @access public
     * @return void
     */
    function __construct(
        $config             = array(),
        $config_storage     = array(),
        $config_lexer       = array(),
        $config_degenerator = array()
    )
    {
        # Validate config data
        foreach ($config as $name => $value) {
            switch ($name) {
                case 'min_dev':
                case 'rob_s':
                case 'rob_x':
                    $this->config[$name] = (float) $value;
                    break;
                case 'use_relevant':
                    $this->config[$name] = (int) $value;
                    break;
                case 'lexer':
                case 'degenerator':
                case 'storage':
                    $this->config[$name] = (string) $value;
                    break;
                default:
                    throw new Exception("b8: Unknown configuration key: \"$name\"");
            }
        }

        # Setup the degenerator class
        $class = $this->_loadClass('degenerator', $this->config['degenerator']);
        if ($class === false) {
            throw new Exception(
                "b8: Could not load class definition file for degenerator " .
                "\"{$this->config['degenerator']}\""
            );
        }
        $this->degenerator = new $class($config_degenerator);

        # Setup the lexer class
        $class = $this->_loadClass('lexer', $this->config['lexer']);
        if ($class === false) {
            throw new Exception(
                "b8: Could not load class definition file for lexer \"{$this->config['lexer']}\""
            );
        }
        $this->lexer = new $class($config_lexer);

        # Setup the storage backend
        $class = $this->_loadClass('storage', 'base');
        if ($class === false) {
            throw new Exception("b8: Could not load class definition file for the storage base class");
        }
        $class = $this->_loadClass('storage', $this->config['storage']);
        if ($class === false) {
            throw new Exception(
                "b8: Could not load class definition file for storage backend " .
                "\"{$this->config['storage']}\""
            );
        }
        $this->storage = new $class($config_storage, $this->degenerator);
    }

    /**
     * Load a class file if a class has not been defined yet.
     *
     * @access private
     * @return boolean Returns true if everything is okay, otherwise false.
     */
    private function _loadClass($class_type, $class_name)
    {
        $complete_class_name = "b8_{$class_type}_{$class_name}";
        $class_file = dirname(__FILE__) . DIRECTORY_SEPARATOR . $class_type . DIRECTORY_SEPARATOR . "{$class_type}_{$class_name}.php";

        if (class_exists($complete_class_name, false) === false) {
            # Check if the requested file actually exists
            if (is_file($class_file) !== true or is_readable($class_file) !== true) {
                return false;
            }

            # Include it
            $included = require_once($class_file);

            if ($included === false or class_exists($complete_class_name, false) === false) {
                return false;
            }
        }

        return $complete_class_name;
    }

    /**
     * Classifies a text
     *
     * @access public
     * @param string $text
     * @return mixed float The rating between 0 (ham) and 1 (spam) or an error code
     */
    public function classify($text = null)
    {
        # Let's first see if the user called the function correctly
        if ($text === null) {
            return self::CLASSIFYER_TEXT_MISSING;
        }

        # Get the internal database variables, containing the number of ham and
        # spam texts so the spam probability can be calculated in relation to them
        $internals = $this->storage->getInternals();

        # Calculate the spamminess of all tokens

        # Get all tokens we want to rate
        $tokens = $this->lexer->getTokens($text);

        # Check if the lexer failed
        # (if so, $tokens will be a lexer error code, if not, $tokens will be an array)
        if (!is_array($tokens)) {
            return $tokens;
        }

        # Fetch all available data for the token set from the database
        $this->_token_data = $this->storage->get(array_keys($tokens));

        # Calculate the spamminess and importance for each token (or a degenerated form of it)

        $word_count = array();
        $rating     = array();
        $importance = array();

        foreach ($tokens as $word => $count) {
            $word_count[$word] = $count;

            # Although we only call this function only here ... let's do the
            # calculation stuff in a function to make this a bit less confusing ;-)
            $rating[$word] = $this->_getProbability(
                $word, $internals['texts_ham'], $internals['texts_spam']
            );

            $importance[$word] = abs(0.5 - $rating[$word]);
        }

        # Order by importance
        arsort($importance);
        reset($importance);

        # Get the most interesting tokens (use all if we have less than the given number)
        $relevant = array();
        for ($i = 0; $i < $this->config['use_relevant']; $i++) {
            if ($token = key($importance)) {
                # Important tokens remain

                # If the token's rating is relevant enough, use it
                if (abs(0.5 - $rating[$token]) > $this->config['min_dev']) {
                    # Tokens that appear more than once also count more than once
                    for ($x = 0, $l = $word_count[$token]; $x < $l; $x++) {
                        array_push($relevant, $rating[$token]);
                    }
                }
            } else {
                # We have less words as we want to use, so we
                # already use what we have and can break here
                break;
            }

            next($importance);
        }

        # Calculate the spamminess of the text (thanks to Mr. Robinson ;-)
        # We set both hamminess and spamminess to 1 for the first multiplying
        $hamminess  = 1;
        $spamminess = 1;

        # Consider all relevant ratings
        foreach ($relevant as $value) {
            $hamminess  *= (1.0 - $value);
            $spamminess *= $value;
        }

        # If no token was good for calculation, we really don't know how
        # to rate this text, so can return 0.5 without further calculations.
        if ($hamminess == 1 and $spamminess == 1) {
            return 0.5;
        }

        # Calculate the combined rating

        # Get the number of relevant ratings
        $n = count($relevant);

        # The actual hamminess and spamminess
        $hamminess  = 1 - pow($hamminess,  (1 / $n));
        $spamminess = 1 - pow($spamminess, (1 / $n));

        # Calculate the combined indicator
        $probability = ($hamminess - $spamminess) / ($hamminess + $spamminess);

        # We want a value between 0 and 1, not between -1 and +1, so ...
        $probability = (1 + $probability) / 2;

        # Alea iacta est
        return $probability;
    }

    /**
     * Calculate the spamminess of a single token also considering "degenerated" versions
     *
     * @access private
     * @param string $word
     * @param string $texts_ham
     * @param string $texts_spam
     * @return void
     */
    private function _getProbability($word, $texts_ham, $texts_spam)
    {
        # Let's see what we have!
        if (isset($this->_token_data['tokens'][$word]) === true) {
            # The token is in the database, so we can use it's data as-is
            # and calculate the spamminess of this token directly
            return $this->_calcProbability(
                $this->_token_data['tokens'][$word], $texts_ham, $texts_spam
            );
        }

        # The token was not found, so do we at least have similar words?
        if (isset($this->_token_data['degenerates'][$word]) === true) {
            # We found similar words, so calculate the spamminess for each one
            # and choose the most important one for the further calculation

            # The default rating is 0.5 simply saying nothing
            $rating = 0.5;

            foreach ($this->_token_data['degenerates'][$word] as $degenerate => $count) {
                # Calculate the rating of the current degenerated token
                $rating_tmp = $this->_calcProbability($count, $texts_ham, $texts_spam);

                # Is it more important than the rating of another degenerated version?
                if(abs(0.5 - $rating_tmp) > abs(0.5 - $rating)) {
                    $rating = $rating_tmp;
                }
            }

            return $rating;
        } else {
            # The token is really unknown, so choose the default rating
            # for completely unknown tokens. This strips down to the
            # robX parameter so we can cheap out the freaky math ;-)
            return $this->config['rob_x'];
        }
    }

    /**
     * Do the actual spamminess calculation of a single token
     *
     * @access private
     * @param array $data
     * @param string $texts_ham
     * @param string $texts_spam
     * @return void
     */
    private function _calcProbability($data, $texts_ham, $texts_spam)
    {
        # Calculate the basic probability as proposed by Mr. Graham

        # But: consider the number of ham and spam texts saved instead of the
        # number of entries where the token appeared to calculate a relative
        # spamminess because we count tokens appearing multiple times not just
        # once but as often as they appear in the learned texts.

        $rel_ham = $data['count_ham'];
        $rel_spam = $data['count_spam'];

        if ($texts_ham > 0) {
            $rel_ham = $data['count_ham'] / $texts_ham;
        }

        if ($texts_spam > 0) {
            $rel_spam = $data['count_spam'] / $texts_spam;
        }

        $rating = $rel_spam / ($rel_ham + $rel_spam);

        # Calculate the better probability proposed by Mr. Robinson
        $all = $data['count_ham'] + $data['count_spam'];
        return (($this->config['rob_s'] * $this->config['rob_x']) + ($all * $rating)) /
               ($this->config['rob_s'] + $all);
    }

    /**
     * Check the validity of the category of a request
     *
     * @access private
     * @param string $category
     * @return void
     */
    private function _checkCategory($category)
    {
        return $category === self::HAM or $category === self::SPAM;
    }

    /**
     * Learn a reference text
     *
     * @access public
     * @param string $text
     * @param const $category Either b8::SPAM or b8::HAM
     * @return mixed void or an error code
     */
    public function learn($text = null, $category = null)
    {
        # Let's first see if the user called the function correctly
        if ($text === null) {
            return self::TRAINER_TEXT_MISSING;
        }
        if ($category === null) {
            return self::TRAINER_CATEGORY_MISSING;
        }

        return $this->_processText($text, $category, self::LEARN);
    }

    /**
     * Unlearn a reference text
     *
     * @access public
     * @param string $text
     * @param const $category Either b8::SPAM or b8::HAM
     * @return mixed void or an error code
     */
    public function unlearn($text = null, $category = null)
    {
        # Let's first see if the user called the function correctly
        if ($text === null) {
            return self::TRAINER_TEXT_MISSING;
        }
        if ($category === null) {
            return self::TRAINER_CATEGORY_MISSING;
        }

        return $this->_processText($text, $category, self::UNLEARN);
    }

    /**
     * Does the actual interaction with the storage backend for learning or unlearning texts
     *
     * @access private
     * @param string $text
     * @param const $category Either b8::SPAM or b8::HAM
     * @param const $action Either b8::LEARN or b8::UNLEARN
     * @return mixed void or an error code
     */
    private function _processText($text, $category, $action)
    {
        # Look if the request is okay
        if ($this->_checkCategory($category) === false) {
            return self::TRAINER_CATEGORY_FAIL;
        }

        # Get all tokens from $text
        $tokens = $this->lexer->getTokens($text);

        # Check if the lexer failed
        # (if so, $tokens will be a lexer error code, if not, $tokens will be an array)
        if (!is_array($tokens)) {
            return $tokens;
        }

        # Pass the tokens and what to do with it to the storage backend
        return $this->storage->processText($tokens, $category, $action);
    }

}

?>